references/debugging_guide.md

```diff
@@ -0,0 +1,95 @@
+# Expected Debugging Protocol [Work-In-Progress]
+
+Based loosely on material from:
+- Zeller, Andreas. Why programs fail: a guide to systematic debugging. Elsevier, 2009.
+- Agans, David J. Debugging: The 9 indispensable rules for finding even the most elusive software and hardware problems. Amacom, 2002.
+
+All suggestions for additional sources of useful info are welcome!
+
+Briefly, a word on words used in this document:
+
+"Bug" is used to generally denote the state of something being broken or incorrect. A "failure" is a bug in the observed behavior of one's code. An "infection" is a bug in the state of a particular object or variable. And finally, a "defect" is a bug in one's code. Generally, one or more defects is written into one's code, which, combined with particular program inputs, causes one or more infections that leads to a failure.
+
+To find and fix the pesky bugs, follow the steps below.
+
+1. Track the failure.
+   1. Save one's work and the current state of the system. Check in the saved state via git with a descriptive message using the following format "[Bug] Insert 73-character-or-less bug description here."
+   2. Create a github issue that documents the bug's existence. This will be a bug report containing:
+      - A one line description of the problem. Should basically be the same as the message from the previous step.
+      - A written description of the failure. What happened, in words?
+      - A written description of the expected behavior. What was supposed to happen, in words?
+      - Severity categorization: {Blocker; Major - aka large problem; Minor - aka an easy to fix problem; Enhancement - aka not a requirement but nice to have.}
+      - A minimal working example or set of steps that reproduces the failure.
+      - If possible (but not required now), a simplified test example that produces the same qualitative failure.
+      - Any diagnostic information (errors, execution logs, etc.) reported when the failure was observed.
+      - Environment information when the failure was observed (installed packages, operating system, memory capacity, etc.)
+   3. Create a new branch off the feature branch to track work on fixing the bug.
+   4. Open a pull request for the new branch to the original feature branch. Register that the issue is being worked on in this pull request by referencing the issue for the bug in the text body of the pull request on github.
+   5. Add a test to the suite of regression tests to reproduce this specific software failure instance. Check the test in with git.
+      - See chapter 4, "Reproducing Problems" of "Why Programs Fail: A Guide to Systematic Debugging" by Morgan Kaufmann.
+   6. Create an automated and simplified test case for this specific failure instance.
+      - See chapter 5, "Simplifying Problems" of "Why Programs Fail: A Guide to Systematic Debugging" by Morgan Kaufmann.
```

Comment:
@timothyb0912 Wouldn't these steps be needed only if the following step is taken?

> - If possible (but not required now), a simplified test example that produces the same qualitative failure.

Reply:
Edited to note the difference between the simplified test and the regression test.